v1/tarball/image.go

```diff
@@ -95,69 +100,83 @@ func (i *image) Manifest() (*v1.Manifest, error) {
 	return i.manifest, nil
 }
 
-// singleManifest is the struct used inside the `manifest.json` file of a `docker save` tarball.
-type singleManifest struct {
+// singleImageTarDescriptor is the struct used to represent a single image inside a `docker save` tarball.
+type singleImageTarDescriptor struct {
 	Config   string
 	RepoTags []string
 	Layers   []string
 }
 
-// tarManifest is the struct used to represent a single image inside a `docker save` tarball.
-type tarManifest []singleManifest
+// tarDescriptor is the struct used inside the `manifest.json` file of a `docker save` tarball.
+type tarDescriptor []singleImageTarDescriptor
 
-func findSpecifiedManifest(tag *name.Tag, tarMfst tarManifest) (*singleManifest, error) {
+func (td tarDescriptor) findSpecifiedImageDescriptor(tag *name.Tag) (*singleImageTarDescriptor, error) {
 	if tag == nil {
-		if len(tarMfst) != 1 {
+		if len(td) != 1 {
 			return nil, errors.New("tarball must contain only a single image to be used with tarball.Image")
 		}
-		return &tarMfst[0], nil
+		return &(td)[0], nil
 	}
-	for _, mfst := range tarMfst {
-		for _, tagStr := range mfst.RepoTags {
+	for _, img := range td {
+		for _, tagStr := range img.RepoTags {
 			repoTag, err := name.NewTag(tagStr, name.WeakValidation)
 			if err != nil {
 				return nil, err
 			}
 
 			// Compare the resolved names, since there are several ways to specify the same tag.
 			if repoTag.Name() == tag.Name() {
-				return &mfst, nil
+				return &img, nil
 			}
 		}
 	}
 	return nil, fmt.Errorf("tag %s not found in tarball", tag)
 }
 
-func (i *image) loadManifest() error {
-	i.lock.Lock()
-	defer i.lock.Unlock()
-
-	mfstBytes, err := extractFileFromTar(i.path, "manifest.json")
+func (i *image) loadTarDescriptorAndConfig() error {
```

Comment:
Why not do this eagerly and avoid the need for locking?

Reply:
Do you mean load these during the construction of the "Image" type in func Image()...?

Reply:
Yeah.